llm/server.go

```diff
@@ -217,15 +218,49 @@ func NewLlamaServer(gpus discover.GpuInfoList, model string, ggml *GGML, adapter
 		params = append(params, "--threads", strconv.Itoa(defaultThreads))
 	}
 
-	flashAttnEnabled := envconfig.FlashAttention()
+	// Check flash attention support
+	flashAttnRequested := envconfig.FlashAttention()
+	if flashAttnRequested {
+		if gpus.SupportsFlashAttention() && supportsFlashAttention(ggml) {
+			params = append(params, "--flash-attn")
+			slog.Info("Enabling flash attention")
 
-	for _, g := range gpus {
-		// only cuda (compute capability 7+) and metal support flash attention
-		if g.Library != "metal" && (g.Library != "cuda" || g.DriverMajor < 7) {
-			flashAttnEnabled = false
+			// Only validate and set KV cache type when flash attention is enabled
+			kvCacheType := envconfig.KvCacheType()
+			if kvCacheType != "" {
+				isEmbeddingModel := false
+				if _, ok := ggml.KV()[fmt.Sprintf("%s.pooling_type", ggml.KV().Architecture())]; ok {
+					isEmbeddingModel = true
+				}
+
+				if validatedType, _ := ValidateKVCacheType(kvCacheType, isEmbeddingModel); validatedType != "" {
+					params = append(params, "--kv-cache-type", validatedType)
+					slog.Debug("Setting cache type", "type", validatedType)
+				}
+			}
```

Comment:
I think we can reduce this a little more to something like:

```
if !slices.Contains(ValidKVCacheTypes, cacheType) {
    return nil, errors.New("invalid k/v cache type")
}

var isEmbeddingModel
if _, ok := ggml.KV()[fmt.Sprintf("%s.pooling_type", ggml.KV().Architecture())]; ok {
    isEmbeddingModel = true
    kvCacheType = "f16"
}

params = append(params, "--kv-cache-type", validatedType)
```

Reply:
Updated.